### PR TITLE
Fix Servo Driver and Add Calibration

### PR DESCRIPTION
**Pull request by masonturner**

### CODE REVIEW

File: embedded/servo_driver/cfg/servo_configs.yaml [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
+  calibrated_min_pwm_count:  1328
+  max_angle_radians: 0.174533
 3:
   index: 3
-  calibrated_center: 50
-  calibrated_max: 71
-  max_angle: 20
+  calibrated_max_pwm_count:  2739
+  calibrated_min_pwm_count:  1257
+  max_angle_radians: 0.174533
```
Comment (jpanikulam) [inline embedded/servo_driver/cfg/servo_configs.yaml]: Good. Job.

File: embedded/servo_driver/servo_calibrate.cc [hunk 2]
```diff
@@ -41,9 +41,9 @@ int keypress(int count, int key) {
   ServoDriver &servo = servos[(num - 1) / 2];
   Direction direction = num % 2 == 1 ? UP : DOWN;
   if (direction == DOWN) {
-    servo.set_percentage(std::max(0.0f, servo.get_percentage() - 1.0f));
+    servo.set_percentage(std::max(0.0, servo.get_percentage() - 0.005));
   } else {
-    servo.set_percentage(std::min(100.0f, servo.get_percentage() + 1.0f));
+    servo.set_percentage(std::min(1.0, servo.get_percentage() + 0.005));
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_calibrate.cc]: What's the +0.005?

Consider making everything `double` and removing the `_f` suffix. It's very inconsistent in the code

File: embedded/servo_driver/servo_driver.cc [hunk 3]
```diff
@@ -32,46 +22,48 @@ ServoDriver::ServoDriver(const Config& config) {
...
   }
 
-  percentage_ = calibrated_center_;
-  set_percentage(percentage_);
+  set_angle_radians(0);
 }
 
-void ServoDriver::set_percentage(float unchecked_percentage) {
-  if (unchecked_percentage > 100 || unchecked_percentage < 0) {
+void ServoDriver::set_percentage(double unchecked_percentage, uint max_pwm_count, uint min_pwm_count) {
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: Make the arguments const

Comment (jpanikulam) [reply embedded/servo_driver/servo_driver.cc]: Specify the number of bits (uint8_t, uint32_t, uint64_t?)

File: embedded/servo_driver/servo_driver.cc [hunk 4]
```diff
@@ -32,46 +22,48 @@ ServoDriver::ServoDriver(const Config& config) {
...
+  if (unchecked_percentage > 1 || unchecked_percentage < 0) {
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: Make this 1.0 and 0.0

File: embedded/servo_driver/servo_driver.cc [hunk 5]
```diff
@@ -32,46 +22,48 @@ ServoDriver::ServoDriver(const Config& config) {
...
 }
 
-void ServoDriver::set_angle_radians(
-    float angle_radians) {  // TODO noble to give units to angle
-  float angle_degrees = rad_to_deg(angle_radians);
-  float angleFraction = static_cast<float>(angle_degrees) / max_angle_;
-  float half_range = (calibrated_max_ - calibrated_center_);
-  set_percentage(calibrated_center_ + half_range * angleFraction);
+void ServoDriver::set_angle_radians(double angle_radians) {
+  const double slope = 1.0 / (2*max_angle_radians_);
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: 2 -> 2.0

File: embedded/servo_driver/servo_driver.cc [hunk 6]
```diff
@@ -32,46 +22,48 @@ ServoDriver::ServoDriver(const Config& config) {
...
+  const double percentage = slope * (angle_radians + max_angle_radians_);
+  set_percentage(percentage, calibrated_max_pwm_count_, calibrated_min_pwm_count_);
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: Add comments clarifying how to use this. We expect angles to be able to be negative?

File: embedded/servo_driver/servo_driver.hh [hunk 7]
```diff
@@ -17,20 +17,29 @@
...
 
  private:
+
+  // 850 to 2150 microseconds for the PWM width, specified by the servo
+  // Values checked on Oscilloscope
+  static constexpr uint MIN_PWM_COUNTS = 1160;
+  static constexpr uint MAX_PWM_COUNTS = 2935;
+  static constexpr uint PWM_FREQUENCY = 330;
   int servo_index_ = -1;
   std::shared_ptr<PwmDriver> pwm_driver_;
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.hh]: [nit] Consider making this a `std::unique_ptr<>`

**Comment by masonturner** [comment]
Added a publisher for commanded pwm set points

**Pull request merged**